compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp

```diff
@@ -405,19 +415,127 @@ CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
+      consumerInfo.getTransformedCollapsableLoops(
+          consumerToProducerMap.value());
+  if (failed(consumerCollapsable)) {
+    didChange = collapsableLoops.size();
+    this->clear();
+    return didChange;
+  }
+
+  FailureOr<SmallVector<ReassociationIndices>> consumerReassoc =
+      consumerInfo.getTransformedReassociation(consumerToProducerMap.value());
+  assert(succeeded(consumerReassoc));
+
+  // Get a map from original index to the index it gets collapsed into
+  llvm::DenseMap<long, long> consumerCollapseMap;
+  for (const auto &[idx, indicies] : llvm::enumerate(consumerReassoc.value())) {
+    for (const auto elem : indicies) {
+      consumerCollapseMap[elem] = idx;
+    }
+  }
+
+  // Remove all collapsable loops in `producer` that are not collapsable in
+  // `consumer` (set intersect)
+  didChange = collapsableLoops.remove_if([&](long elem) -> bool {
+    return !consumerCollapsable.value().contains(elem);
+  });
+
+  // Now update the reassociation indicies given the updated `collapsableLoops`
+  // and `consumerCollapsableMap`.
+  // The idea is to reconstruct the reassociation indicies, and at each index:
+  // (1) If `index` IS NOT in `collapsableLoops`, split `indicies` and don't add
+  // `index` to either.
+  //
+  // (2) If `index` IS in `collapsableLoops` but `consumerCollapseMap` maps
+  // `index` to a different collapsed loop then the other indicies,  split
+  // `indicies` and insert `index` into the new one.
+  //
+  // For example:
+  // producer reassociation = [[0, 1], [2, 3]]
+  // consumer reassociation = [0, 1, 2, 3]
+  // then, consumer reassociation gets updated to [[0, 1], [2, 3]] because
+  // [0, 1] and [2, 3] get collapsed into different loops
+  //
+  // (3) Otherwise, keep the index
+  constexpr long kUninitialized = -1;
+  SmallVector<ReassociationIndices> newReassociation;
+  for (ReassociationIndicesRef indicies : reassociation) {
+    // Track the loop index that `indicies` get collapsed into.
+    long collapseIntoIdx = kUninitialized;
+
+    // Holds dimensions that should be collapsed together
+    ReassociationIndices newIndicies;
+    for (int64_t index : indicies) {
+      // (1) Because `index` isn't collapsable, the indicies in `newIndicies`
+      // are no longer adjacent to the upcoming indicies. If there is >1 index
+      // to collapse, add it to the new reassociation. Otherwise, discard it
+      // because there is no dimension to collapse with.
+      if (!collapsableLoops.contains(index)) {
+        if (newIndicies.size() > 1) {
+          newReassociation.push_back(std::move(newIndicies));
+        }
+        newIndicies.clear();
+        collapseIntoIdx = kUninitialized;
+        continue;
+      }
+
+      if (collapseIntoIdx == kUninitialized) {
+        collapseIntoIdx = consumerCollapseMap.at(index);
+      }
+
+      if (collapseIntoIdx != kUninitialized &&
+          consumerCollapseMap.at(index) != collapseIntoIdx) {
```

Comment:
If I read it correctly, can we just use `else` here?

```suggestion
      } else {
```